### ISSUE TITLE
NetworkManager fails using internal DHCP client

### ISSUE THREAD

**Issue by jake-dog** [pop-os/pop#2904]
**Distribution (run `cat /etc/os-release`):**
 
```
NAME="Pop!_OS"
VERSION="22.04 LTS"
ID=pop
ID_LIKE="ubuntu debian"
PRETTY_NAME="Pop!_OS 22.04 LTS"
VERSION_ID="22.04"
HOME_URL="https://pop.system76.com"
SUPPORT_URL="https://support.system76.com"
BUG_REPORT_URL="https://github.com/pop-os/pop/issues"
PRIVACY_POLICY_URL="https://system76.com/privacy"
VERSION_CODENAME=jammy
UBUNTU_CODENAME=jammy
LOGO=distributor-logo-pop-os
```

```
$ uname -a
Linux hpdevone 6.2.6-76060206-generic #202303130630~1683753207~22.04~77c1465 SMP PREEMPT_DYNAMIC Wed M x86_64 x86_64 x86_64 GNU/Linux
```

```
$ lspci -nnk -s 01:00.0
01:00.0 Network controller [0280]: Intel Corporation Wi-Fi 6 AX210/AX211/AX411 160MHz [8086:2725] (rev 1a)
	DeviceName: WLAN
	Subsystem: Intel Corporation Wi-Fi 6 AX210 160MHz [8086:0024]
	Kernel driver in use: iwlwifi
	Kernel modules: iwlwifi
```

**Related Application and/or Package Version (run `apt policy $PACKAGE NAME`):**

```
network-manager:
  Installed: 1.36.6-0ubuntu2
  Candidate: 1.36.6-0ubuntu2
  Version table:
 *** 1.36.6-0ubuntu2 500
        500 http://apt.pop-os.org/ubuntu jammy-updates/main amd64 Packages
        100 /var/lib/dpkg/status
     1.36.4-2ubuntu1 500
        500 http://apt.pop-os.org/ubuntu jammy/main amd64 Packages
```

**Issue/Bug Description:**

Trying to connect to any network, wired or wifi, fails when using network-manager ("NetworkManager"), but works fine when connecting [manually via dhclient/wpa_supplicant/ip](#appendix-1-manual-connect-workaround).  The error logged by network-manager is cryptic: `device (wlo1): Activation: failed for connection 'REDACTED_WIFI_SSID'` (see [appendix 2](#appendix-2---logs-when-trying-to-connect-via-networkmanager) for full logs).

Logs show that wpa_supplicant successfully negotiates with the AP, but whatever network-manager does thereafter fails.  Presumably after WPA negotiation succeeds, DHCP discovery begins.  From tcpdump output it appears that the network-manager's ["internal" DHCP client](https://manpages.ubuntu.com/manpages/bionic/man5/NetworkManager.conf.5.html) is sending DHCP "Discover" packets which never receive an "Offer"-- but thats not the only problem.  (I am not an expert in DHCP and I do not understand why there is no "Offer" response to this packet. See [appendix 5](#appendix-5---successful-dhclient-packet-from-manual-connect) to compare with a successful dhclient generated DHCP "Discover".)

```
17:20:06.933141 IP (tos 0xc0, ttl 64, id 0, offset 0, flags [DF], proto UDP (17), length 314)
    0.0.0.0.68 > 255.255.255.255.67: [udp sum ok] BOOTP/DHCP, Request from REDACTED_MAC_ADDRESS length 286, xid 0x550bbbc9, secs 30, Flags [none] (0x0000)
	  Client-Ethernet-Address REDACTED_MAC_ADDRESS
	  Vendor-rfc1048 Extensions
	    Magic Cookie 0x63825363
	    DHCP-Message (53), length 1: Discover
	    Client-ID (61), length 7: ether REDACTED_MAC_ADDRESS
	    Parameter-Request (55), length 17:
	      Subnet-Mask (1), Time-Zone (2), Domain-Name-Server (6), Hostname (12)
	      Domain-Name (15), MTU (26), BR (28), Classless-Static-Route (121)
	      Default-Gateway (3), Static-Route (33), YD (40), YS (41)
	      NTP (42), Unknown (119), Classless-Static-Route-Microsoft (249), Unknown (252)
	      RP (17)
	    MSZ (57), length 2: 576
	    Hostname (12), length 8: "hpdevone"
	    END (255), length 0
```

Switching [network-manager to use dhclient instead of the internal DHCP client](#appendix-4---configuring-networkmanager-to-use-dhclient) fixes the DHCP issue, but "activation" (whatever that is) still fails in a loop of cryptic errors.  Every few seconds it successfully connects, but then logs the following message, disconnects, reconnects, and so on:

```
May 26 17:55:18 hpdevone NetworkManager[5777]: <warn>  [1685138118.8214] device (wlo1): link timed out.
May 26 17:55:18 hpdevone NetworkManager[5777]: <info>  [1685138118.8221] device (wlo1): state change: activated -> failed (reason 'ssid-not-found', sys-iface-state: 'managed')
May 26 17:55:18 hpdevone NetworkManager[5777]: <info>  [1685138118.8226] manager: NetworkManager state is now DISCONNECTED
May 26 17:55:18 hpdevone NetworkManager[5777]: <warn>  [1685138118.8233] device (wlo1): Activation: failed for connection 'REDACTED_WIFI_SSID'
```

**Steps to reproduce (if you know):**

All attempts to connect to any network via network-manager (drop-down in the top menu) fail.  I do not understand what the cause is and regret to say that I cannot provide a proper reproduction.  I hope that I have provided enough data for resolvers and users facing the same issue.

**Expected behavior:**

Connecting via network-manager should just work, without cryptic errors or workarounds.

**Other Notes:**

Mostly stock [HP dev one](https://hpdevone.com/), with two aftermarket hardware customizations:

* Intel AX210NGW wifi
* Anker (A8313) 1gbps USB-C to ethernet adapter

Power management for wifi is disabled, because it behaves oddly when enabled.

---

## Appendix 1 - Manual connect workaround

I can connect to wifi via the following manual procedure:

1. "Turn off" wifi via the task bar wifi menu
2. `rfkill unblock wlan`
3. `sudo wpa_supplicant -Dwext -i wlo1 -c /tmp/test.conf`
4. `sudo dhclient wlo1`

And voila, just like that I'm connected to the wifi.  (The file `/tmp/test.conf` was generated via the command `wpa_passphrase "REDACTED_WIFI_SSID" > /tmp/test.conf`.)

## Appendix 2 - Logs when trying to connect via NetworkManager

```
May 26 16:17:09 hpdevone NetworkManager[1044]: <info>  [1685132229.4262] device (wlo1): Activation: starting connection 'REDACTED_WIFI_SSID' (585e2ddc-f41c-40ce-81ea-81cbd15b30b3)
May 26 16:17:09 hpdevone NetworkManager[1044]: <info>  [1685132229.4263] audit: op="connection-activate" uuid="585e2ddc-f41c-40ce-81ea-81cbd15b30b3" name="REDACTED_WIFI_SSID" pid=1916 uid=1000 result="success"
May 26 16:17:09 hpdevone NetworkManager[1044]: <info>  [1685132229.4264] device (wlo1): state change: disconnected -> prepare (reason 'none', sys-iface-state: 'managed')
May 26 16:17:09 hpdevone NetworkManager[1044]: <info>  [1685132229.4265] manager: NetworkManager state is now CONNECTING
May 26 16:17:09 hpdevone NetworkManager[1044]: <info>  [1685132229.4268] device (wlo1): state change: prepare -> config (reason 'none', sys-iface-state: 'managed')
May 26 16:17:09 hpdevone NetworkManager[1044]: <info>  [1685132229.4272] device (wlo1): Activation: (wifi) access point 'REDACTED_WIFI_SSID' has security, but secrets are required.
May 26 16:17:09 hpdevone NetworkManager[1044]: <info>  [1685132229.4272] device (wlo1): state change: config -> need-auth (reason 'none', sys-iface-state: 'managed')
May 26 16:17:09 hpdevone NetworkManager[1044]: <info>  [1685132229.4281] device (wlo1): state change: need-auth -> prepare (reason 'none', sys-iface-state: 'managed')
May 26 16:17:09 hpdevone NetworkManager[1044]: <info>  [1685132229.4284] device (wlo1): state change: prepare -> config (reason 'none', sys-iface-state: 'managed')
May 26 16:17:09 hpdevone NetworkManager[1044]: <info>  [1685132229.4286] device (wlo1): Activation: (wifi) connection 'REDACTED_WIFI_SSID' has security, and secrets exist.  No new secrets needed.
May 26 16:17:09 hpdevone NetworkManager[1044]: <info>  [1685132229.4286] Config: added 'ssid' value 'REDACTED_WIFI_SSID'
May 26 16:17:09 hpdevone NetworkManager[1044]: <info>  [1685132229.4286] Config: added 'scan_ssid' value '1'
May 26 16:17:09 hpdevone NetworkManager[1044]: <info>  [1685132229.4286] Config: added 'bgscan' value 'simple:30:-70:86400'
May 26 16:17:09 hpdevone NetworkManager[1044]: <info>  [1685132229.4287] Config: added 'key_mgmt' value 'WPA-PSK WPA-PSK-SHA256 FT-PSK SAE FT-SAE'
May 26 16:17:09 hpdevone NetworkManager[1044]: <info>  [1685132229.4287] Config: added 'psk' value '<hidden>'
May 26 16:17:09 hpdevone NetworkManager[1044]: <info>  [1685132229.4398] device (wlo1): supplicant interface state: disconnected -> scanning
May 26 16:17:09 hpdevone NetworkManager[1044]: <info>  [1685132229.4399] device (p2p-dev-wlo1): supplicant management interface state: disconnected -> scanning
May 26 16:17:11 hpdevone wpa_supplicant[912]: wlo1: SME: Trying to authenticate with REDACTED_WIFI_MAC_ADDRESS (SSID='REDACTED_WIFI_SSID' freq=2452 MHz)
May 26 16:17:11 hpdevone kernel: wlo1: authenticate with REDACTED_WIFI_MAC_ADDRESS
May 26 16:17:11 hpdevone kernel: wlo1: 80 MHz not supported, disabling VHT
May 26 16:17:11 hpdevone kernel: wlo1: send auth to REDACTED_WIFI_MAC_ADDRESS (try 1/3)
May 26 16:17:11 hpdevone NetworkManager[1044]: <info>  [1685132231.0634] device (wlo1): supplicant interface state: scanning -> authenticating
May 26 16:17:11 hpdevone NetworkManager[1044]: <info>  [1685132231.0634] device (p2p-dev-wlo1): supplicant management interface state: scanning -> authenticating
May 26 16:17:11 hpdevone wpa_supplicant[912]: wlo1: Trying to associate with REDACTED_WIFI_MAC_ADDRESS (SSID='REDACTED_WIFI_SSID' freq=2452 MHz)
May 26 16:17:11 hpdevone NetworkManager[1044]: <info>  [1685132231.0921] device (wlo1): supplicant interface state: authenticating -> associating
May 26 16:17:11 hpdevone NetworkManager[1044]: <info>  [1685132231.0922] device (p2p-dev-wlo1): supplicant management interface state: authenticating -> associating
May 26 16:17:11 hpdevone kernel: wlo1: authenticated
May 26 16:17:11 hpdevone kernel: wlo1: associate with REDACTED_WIFI_MAC_ADDRESS (try 1/3)
May 26 16:17:11 hpdevone wpa_supplicant[912]: RRM: Ignoring radio measurement request: Not associated
May 26 16:17:11 hpdevone kernel: wlo1: RX AssocResp from REDACTED_WIFI_MAC_ADDRESS (capab=0xc31 status=0 aid=1)
May 26 16:17:11 hpdevone wpa_supplicant[912]: wlo1: Associated with REDACTED_WIFI_MAC_ADDRESS
May 26 16:17:11 hpdevone wpa_supplicant[912]: wlo1: CTRL-EVENT-SUBNET-STATUS-UPDATE status=0
May 26 16:17:11 hpdevone kernel: wlo1: associated
May 26 16:17:11 hpdevone NetworkManager[1044]: <info>  [1685132231.1186] device (wlo1): supplicant interface state: associating -> associated
May 26 16:17:11 hpdevone NetworkManager[1044]: <info>  [1685132231.1187] device (p2p-dev-wlo1): supplicant management interface state: associating -> associated
May 26 16:17:11 hpdevone NetworkManager[1044]: <info>  [1685132231.4991] device (wlo1): supplicant interface state: associated -> 4way_handshake
May 26 16:17:11 hpdevone NetworkManager[1044]: <info>  [1685132231.4992] device (p2p-dev-wlo1): supplicant management interface state: associated -> 4way_handshake
May 26 16:17:11 hpdevone wpa_supplicant[912]: wlo1: WPA: Key negotiation completed with REDACTED_WIFI_MAC_ADDRESS [PTK=CCMP GTK=CCMP]
May 26 16:17:11 hpdevone wpa_supplicant[912]: wlo1: CTRL-EVENT-CONNECTED - Connection to REDACTED_WIFI_MAC_ADDRESS completed [id=0 id_str=]
May 26 16:17:11 hpdevone NetworkManager[1044]: <info>  [1685132231.5833] device (wlo1): supplicant interface state: 4way_handshake -> completed
May 26 16:17:11 hpdevone NetworkManager[1044]: <info>  [1685132231.5834] device (wlo1): Activation: (wifi) Stage 2 of 5 (Device Configure) successful. Connected to wireless network "REDACTED_WIFI_SSID"
May 26 16:17:11 hpdevone wpa_supplicant[912]: wlo1: CTRL-EVENT-SIGNAL-CHANGE above=1 signal=-31 noise=9999 txrate=1000
May 26 16:17:11 hpdevone NetworkManager[1044]: <info>  [1685132231.5834] device (p2p-dev-wlo1): supplicant management interface state: 4way_handshake -> completed
May 26 16:17:11 hpdevone NetworkManager[1044]: <info>  [1685132231.5836] device (wlo1): state change: config -> ip-config (reason 'none', sys-iface-state: 'managed')
May 26 16:17:11 hpdevone NetworkManager[1044]: <info>  [1685132231.5841] dhcp4 (wlo1): activation: beginning transaction (timeout in 45 seconds)
May 26 16:17:11 hpdevone avahi-daemon[844]: Joining mDNS multicast group on interface wlo1.IPv6 with address fe80::f23f:7049:7b71:af52.
May 26 16:17:11 hpdevone avahi-daemon[844]: New relevant interface wlo1.IPv6 for mDNS.
May 26 16:17:11 hpdevone avahi-daemon[844]: Registering new address record for fe80::f23f:7049:7b71:af52 on wlo1.*.
May 26 16:17:51 hpdevone systemd[1]: Starting Network Manager Script Dispatcher Service...
May 26 16:17:51 hpdevone systemd[1]: Started Network Manager Script Dispatcher Service.
May 26 16:17:56 hpdevone NetworkManager[1044]: <info>  [1685132276.8230] device (wlo1): state change: ip-config -> failed (reason 'ip-config-unavailable', sys-iface-state: 'managed')
May 26 16:17:56 hpdevone NetworkManager[1044]: <info>  [1685132276.8235] manager: NetworkManager state is now DISCONNECTED
May 26 16:17:56 hpdevone NetworkManager[1044]: <warn>  [1685132276.8239] device (wlo1): Activation: failed for connection 'REDACTED_WIFI_SSID'
May 26 16:17:56 hpdevone NetworkManager[1044]: <info>  [1685132276.8241] device (wlo1): state change: failed -> disconnected (reason 'none', sys-iface-state: 'managed')
May 26 16:17:56 hpdevone avahi-daemon[844]: Withdrawing address record for fe80::f23f:7049:7b71:af52 on wlo1.
May 26 16:17:56 hpdevone avahi-daemon[844]: Leaving mDNS multicast group on interface wlo1.IPv6 with address fe80::f23f:7049:7b71:af52.
May 26 16:17:56 hpdevone avahi-daemon[844]: Interface wlo1.IPv6 no longer relevant for mDNS.
May 26 16:17:56 hpdevone NetworkManager[1044]: <info>  [1685132276.8671] dhcp4 (wlo1): canceled DHCP transaction
May 26 16:17:56 hpdevone kernel: wlo1: deauthenticating from REDACTED_WIFI_MAC_ADDRESS by local choice (Reason: 3=DEAUTH_LEAVING)
May 26 16:17:57 hpdevone wpa_supplicant[912]: wlo1: CTRL-EVENT-DISCONNECTED bssid=REDACTED_WIFI_MAC_ADDRESS reason=3 locally_generated=1
May 26 16:17:57 hpdevone wpa_supplicant[912]: wlo1: CTRL-EVENT-DSCP-POLICY clear_all
May 26 16:17:57 hpdevone NetworkManager[1044]: <info>  [1685132277.1445] device (wlo1): supplicant interface state: completed -> disconnected
May 26 16:17:57 hpdevone NetworkManager[1044]: <info>  [1685132277.1447] device (p2p-dev-wlo1): supplicant management interface state: completed -> disconnected
May 26 16:17:57 hpdevone wpa_supplicant[912]: wlo1: Reject scan trigger since one is already pending
May 26 16:18:03 hpdevone wpa_supplicant[912]: wlo1: Reject scan trigger since one is already pending
May 26 16:18:06 hpdevone systemd[1]: NetworkManager-dispatcher.service: Deactivated successfully.
```

## Appendix 3 - Logs when manually connecting

Logs are from executing the procedure described in [appendix 1](#appendix-1-manual-connect-workaround)

```
May 26 15:35:53 hpdevone NetworkManager[10466]: <info>  [1685129753.1762] manager: rfkill: Wi-Fi hardware radio set disabled
May 26 15:35:53 hpdevone NetworkManager[10466]: <info>  [1685129753.1764] device (wlo1): state change: disconnected -> unavailable (reason 'none', sys-iface-state: 'managed')
May 26 15:35:53 hpdevone wpa_supplicant[18721]: rfkill: WLAN soft blocked
May 26 15:35:53 hpdevone wpa_supplicant[18721]: rfkill: WLAN soft blocked
May 26 15:35:53 hpdevone wpa_supplicant[18721]: p2p-dev-wlo1: CTRL-EVENT-DSCP-POLICY clear_all
May 26 15:35:53 hpdevone wpa_supplicant[18721]: p2p-dev-wlo1: CTRL-EVENT-DSCP-POLICY clear_all
May 26 15:35:53 hpdevone wpa_supplicant[18721]: nl80211: deinit ifname=p2p-dev-wlo1 disabled_11b_rates=0
May 26 15:35:53 hpdevone systemd[1]: Starting Load/Save RF Kill Switch Status...
May 26 15:35:53 hpdevone systemd[1]: Started Load/Save RF Kill Switch Status.
May 26 15:35:53 hpdevone wpa_supplicant[18721]: wlo1: CTRL-EVENT-DSCP-POLICY clear_all
May 26 15:35:53 hpdevone NetworkManager[10466]: <info>  [1685129753.1957] audit: op="radio-control" arg="wireless-enabled:off" pid=2941 uid=1000 result="success"
May 26 15:35:53 hpdevone NetworkManager[10466]: <info>  [1685129753.1959] manager: rfkill: Wi-Fi now disabled by radio killswitch
May 26 15:35:53 hpdevone NetworkManager[10466]: <info>  [1685129753.1960] device (p2p-dev-wlo1): state change: disconnected -> unavailable (reason 'supplicant-failed', sys-iface-state: 'managed')
May 26 15:35:53 hpdevone wpa_supplicant[18721]: wlo1: CTRL-EVENT-DSCP-POLICY clear_all
May 26 15:35:53 hpdevone wpa_supplicant[18721]: nl80211: deinit ifname=wlo1 disabled_11b_rates=0
May 26 15:35:58 hpdevone systemd[1]: systemd-rfkill.service: Deactivated successfully.
May 26 15:36:00 hpdevone rfkill[18981]: unblock set for type wlan
May 26 15:36:00 hpdevone NetworkManager[10466]: <info>  [1685129760.9962] manager: rfkill: Wi-Fi now enabled by radio killswitch
May 26 15:36:01 hpdevone systemd[1]: Starting Load/Save RF Kill Switch Status...
May 26 15:36:01 hpdevone systemd[1]: Started Load/Save RF Kill Switch Status.
May 26 15:36:06 hpdevone systemd[1]: systemd-rfkill.service: Deactivated successfully.
May 26 15:36:13 hpdevone sudo[18999]:     jake : TTY=pts/1 ; PWD=/home/jake ; USER=root ; COMMAND=/usr/sbin/wpa_supplicant -Dwext -i wlo1 -c /tmp/test.conf
May 26 15:36:13 hpdevone sudo[18999]: pam_unix(sudo:session): session opened for user root(uid=0) by (uid=1000)
May 26 15:36:14 hpdevone kernel: iwlwifi 0000:01:00.0: WFPM_UMAC_PD_NOTIFICATION: 0x1f
May 26 15:36:14 hpdevone kernel: iwlwifi 0000:01:00.0: WFPM_LMAC2_PD_NOTIFICATION: 0x1f
May 26 15:36:14 hpdevone kernel: iwlwifi 0000:01:00.0: WFPM_AUTH_KEY_0: 0x80
May 26 15:36:14 hpdevone kernel: iwlwifi 0000:01:00.0: CNVI_SCU_SEQ_DATA_DW9: 0x0
May 26 15:36:17 hpdevone kernel: wlo1: authenticate with REDACTED_WIFI_MAC_ADDRESS
May 26 15:36:17 hpdevone kernel: wlo1: 80 MHz not supported, disabling VHT
May 26 15:36:17 hpdevone kernel: wlo1: send auth to REDACTED_WIFI_MAC_ADDRESS (try 1/3)
May 26 15:36:17 hpdevone kernel: wlo1: send auth to REDACTED_WIFI_MAC_ADDRESS (try 2/3)
May 26 15:36:17 hpdevone kernel: wlo1: authenticated
May 26 15:36:17 hpdevone kernel: wlo1: associate with REDACTED_WIFI_MAC_ADDRESS (try 1/3)
May 26 15:36:17 hpdevone kernel: wlo1: RX AssocResp from REDACTED_WIFI_MAC_ADDRESS (capab=0xc31 status=0 aid=1)
May 26 15:36:17 hpdevone kernel: wlo1: associated
May 26 15:36:18 hpdevone kernel: IPv6: ADDRCONF(NETDEV_CHANGE): wlo1: link becomes ready
May 26 15:36:27 hpdevone sudo[19018]:     jake : TTY=pts/2 ; PWD=/home/jake ; USER=root ; COMMAND=/usr/sbin/dhclient wlo1
May 26 15:36:27 hpdevone sudo[19018]: pam_unix(sudo:session): session opened for user root(uid=0) by (uid=1000)
May 26 15:36:27 hpdevone dhclient[19021]: DHCPDISCOVER on wlo1 to 255.255.255.255 port 67 interval 3 (xid=0xb172f31f)
May 26 15:36:27 hpdevone dhclient[19021]: DHCPOFFER of 10.100.0.25 from 10.100.0.1
May 26 15:36:27 hpdevone dhclient[19021]: DHCPREQUEST for 10.100.0.25 on wlo1 to 255.255.255.255 port 67 (xid=0x1ff372b1)
May 26 15:36:27 hpdevone dhclient[19021]: DHCPACK of 10.100.0.25 from 10.100.0.1 (xid=0xb172f31f)
May 26 15:36:27 hpdevone avahi-daemon[843]: Joining mDNS multicast group on interface wlo1.IPv4 with address 10.100.0.25.
May 26 15:36:27 hpdevone avahi-daemon[843]: New relevant interface wlo1.IPv4 for mDNS.
May 26 15:36:27 hpdevone avahi-daemon[843]: Registering new address record for 10.100.0.25 on wlo1.IPv4.
May 26 15:36:27 hpdevone systemd-resolved[17857]: wlo1: Bus client set search domain list to: home.lab
May 26 15:36:27 hpdevone systemd-resolved[17857]: wlo1: Bus client set DNS server list to: 10.100.0.1
May 26 15:36:27 hpdevone dhclient[19021]: bound to 10.100.0.25 -- renewal in 3262 seconds.
May 26 15:36:27 hpdevone sudo[19018]: pam_unix(sudo:session): session closed for user root
May 26 15:36:27 hpdevone systemd[1]: Starting resolvconf-pull-resolved.service...
May 26 15:36:27 hpdevone sh[19090]: /etc/resolvconf/update.d/libc: Warning: /etc/resolv.conf is not a symbolic link to /run/resolvconf/resolv.conf
May 26 15:36:27 hpdevone systemd[1]: resolvconf-pull-resolved.service: Deactivated successfully.
May 26 15:36:27 hpdevone systemd[1]: Finished resolvconf-pull-resolved.service.
```

wpa_supplicant is launched asynchronously and the command output is shown below:

```
$ sudo wpa_supplicant -Dwext -i wlo1 -c /tmp/test.conf
Successfully initialized wpa_supplicant
ioctl[SIOCSIWENCODEEXT]: Invalid argument
ioctl[SIOCSIWENCODEEXT]: Invalid argument
wlo1: Trying to associate with REDACTED_WIFI_MAC_ADDRESS (SSID='REDACTED_WIFI_SSID' freq=2452 MHz)
wlo1: Associated with REDACTED_WIFI_MAC_ADDRESS
wlo1: WPA: Key negotiation completed with REDACTED_WIFI_MAC_ADDRESS [PTK=CCMP GTK=CCMP]
wlo1: CTRL-EVENT-CONNECTED - Connection to REDACTED_WIFI_MAC_ADDRESS completed [id=0 id_str=]
```

## Appendix 4 - Configuring NetworkManager to use dhclient

```diff
--- /etc/NetworkManager/NetworkManager.conf
+++ /etc/NetworkManager/NetworkManager.conf	2023-05-26 17:52:23.533463977 -0400
@@ -1,5 +1,6 @@
 [main]
 plugins=ifupdown,keyfile
+dhcp=dhclient

 [ifupdown]
 managed=false
```

## Appendix 5 - Successful dhclient packet from manual connect

```
17:34:46.271753 IP (tos 0x10, ttl 128, id 0, offset 0, flags [none], proto UDP (17), length 328)
    0.0.0.0.68 > 255.255.255.255.67: [udp sum ok] BOOTP/DHCP, Request from REDACTED_MAC_ADDRESS, length 300, xid 0x53f1ad4e, Flags [none] (0x0000)
	  Client-Ethernet-Address REDACTED_MAC_ADDRESS
	  Vendor-rfc1048 Extensions
	    Magic Cookie 0x63825363
	    DHCP-Message (53), length 1: Discover
	    Requested-IP (50), length 4: 10.100.0.25
	    Hostname (12), length 8: "hpdevone"
	    Parameter-Request (55), length 13:
	      Subnet-Mask (1), BR (28), Time-Zone (2), Default-Gateway (3)
	      Domain-Name (15), Domain-Name-Server (6), Unknown (119), Hostname (12)
	      Netbios-Name-Server (44), Netbios-Scope (47), MTU (26), Classless-Static-Route (121)
	      NTP (42)
	    END (255), length 0
	    PAD (0), length 0, occurs 25
```

**Comment by jake-dog** [comment]
Switching network-manager's DHCP client to dhcpd (**Appendix 4**)  has seemingly resolved my connectivity issues for both wired and wifi.

Observations from followup research:

1. network-manager [switched to "internal" DHCP client in 2020](https://gitlab.freedesktop.org/NetworkManager/NetworkManager/blob/1.20.0/NEWS#L13)
2. This is a common issue with network-manager's "internal" DHCP client, at least [as far back as 2017](https://forum.salixos.org/viewtopic.php?f=30&t=7284)
3. network-manager seemingly has compatibility issues with common/popular DHCP servers [like openwrt](https://forum.salixos.org/viewtopic.php?f=30&t=7284) and [ISC DHCP](https://www.isc.org/dhcp/)
4. The ["activation failed" error](https://www.google.com/search?q=networkmanager+activation+failed) is ubiquitous and meaningless [_null-pointer error_](https://www.infoq.com/presentations/Null-References-The-Billion-Dollar-Mistake-Tony-Hoare/) for NetworkManager

It can't be understated the importance of painless network connectivity for rubes like me.  Being a layman upon encountering such an obstacle, my first thought was to restart my router and APs.  Searching the internet for the "activation failed" error provided no end of unproductive troubleshooting steps.  Adding insult to injury, I couldn't research the problem on my laptop because the issue I was trying to solve also prevented me from connecting to the network.

While I'm sure the intentions of NetworkManager authors are noble, seven years of instability in such a fundamental component of the desktop experience should raise eyebrows.  I was in the process of abandoning my laptop when I discovered a solution, and I certainly don't want other users to experience a similar inconvenience.

**Comment by jake-dog** [comment]
I created a [small python script](https://gist.github.com/jake-dog/94192e000174de40b2a47ddd3cfa5ccd) to generate an identical DHCP "Discover" packet to the one generated by stock network-manager (v1.36.6).  Limited testing has shown there is nothing wrong with the DHCP packet itself, having sent at least a dozen so far to two different DHCP servers all of which returned, seemingly valid, "Offer" responses.

Given that the DHCP "Discover" itself has been shown to be valid, it seems that the issue must reside in network-manager itself.  Presumably network-manager got itself into a bad state, as evidenced by the failure loop observed after setting `dhcp=dhclient` (**Appendix 4**).

Initial attempts to resolve the issue included reinstalling network-manager, and rebooting.  The Pop_OS [network troubleshooting guide](https://support.system76.com/articles/network/) offers no help for network-manager issues, however the wireless troubleshooting guide includes a section on [useful commands](https://support.system76.com/articles/wireless#useful-commands) with an obscure note:

> After reinstalling the above packages, fully shut down the machine and then power it back on, rather than rebooting. This ensures the hardware completely resets.

Would this have resolved the issue?  If it would have, surely there is a more user friendly way of accomplishing the same without such incantations.
